rfc-157-edge-compute.md

```diff
@@ -0,0 +1,236 @@
+  - Our CloudFront failover distribution is configured and deployed using Terraform, in [the `infra-mirror-bucket` project in `govuk-aws`](https://github.com/alphagov/govuk-aws/tree/main/terraform/projects/infra-mirror-bucket).
+- Currently only our Fastly configuration is feature-complete; if Fastly goes down then we failover to a CloudFront distribution that is only configured to serve a static mirror of GOV.UK.
+  - Platform Reliability have begun work on a new CloudFront distribution that's able to serve content from origin by calling into a load balancer in front of our `cache` class nodes, but this new setup [doesn't have feature parity](https://docs.google.com/document/d/17_dfWvKNmqyLX1h_PPY6_Cd6IggrrSsP-Peh2De6JQk/edit) with our existing VCL-based Fastly service, and we have no plans to implement full feature parity with Fastly (as that's the purpose of this RFC).
+  - This lack of feature parity presents us with several issues:
+    - Lack of clarity or confidence around what works in Cloudfront (e.g. developers have to be aware of this when building things, SMT need to be reminded of impact failover will have, we may be more hesitant in triggering the failover, etc.)
+    - Complexities also introduced downstream (e.g. if we failover and A/B testing stops working, data analysts now need to scrub out the failover period in their analysis)
+    - We're less likely to drill the failover in Production to ensure it works (as it will cause a somewhat degraded experience)
+    - It restricts our ability to consider a multi CDN strategy where traffic is split between two CDNs simultaneously 
+
+## Proposal
+
+Move from VCL-based CDN config to edge compute on both Fastly ([Compute@Edge](https://www.fastly.com/products/edge-compute)) and AWS ([Lambda@Edge](https://aws.amazon.com/lambda/edge/)).
+
+Benefits:
+
+- Allows us to deploy the same CDN config to multiple CDNs
+  - This makes it easier for us to build a failover solution that has feature parity with the primary CDN
+- Using a programming language for CDN config (as opposed to a DSL) allows us to test it locally, as discussed later in this RFC
+
+## Background
+
+### VCL (Fastly)
+
+This is the "traditional" way to configure Fastly services (see [high-level overview](https://developer.fastly.com/learning/vcl/using/)).
+
+[Our VCL](https://github.com/alphagov/govuk-cdn-config/blob/main/vcl_templates/www.vcl.erb) currently overrides the `RECV`, `FETCH`, `DELIVER` and `ERROR` subroutines, and handles a lot of things including (non-exhaustive):
+
+- IP allowlisting
+- Requiring HTTP basic auth on integration
+- A/B testing, including Replatforming's traffic split experiment logic
+- [JA3](https://github.com/salesforce/ja3) fingerprinting and denylisting
+- Enforcing SSL
+- Setting feature flag headers (e.g. `Govuk-Use-Recommended-Related-Links`, which enables recommended links in Whitehall)
```

Comment:
(Slightly off topic) I think we should look at removing/replacing this particular header. It's never been used as a toggle (after launch) and leaks into response headers that get sent to the client. If we want to use feature flags there are probably better ways of doing it now.